@Scheduled method invoked both on original bean and @Alternative bean

### Describe the bug

If `@Scheduled` is applied to a function and an `@Alternative` subclass is created for the bean containing the annotated method, the method is invoked twice by the scheduler, once for each bean. This means that when the alternative bean is being used the scheduled method is invoked twice, which is unexpected.

A reproducer is attached with the minimal changes necessary to show this, applied on the scheduler-quickstart project.

Our use case is we are creating a bean like `CounterBean` in the attached example that lives inside a library.  This `CounterBean` could be subclassed and customized by providing an alternative like `AlternativeCounterBean` in different projects that use the library, and if that is done at the moment we find multiple invocations of the scheduled method, one immediately after the other.

### Expected behavior

Only one invocation of the `@Scheduled` method should be performed on the alternative bean; the original bean's invocation should not be performed.  In the given example, only `AlternativeCounterBean#increment()` should be invoked by the scheduler.

### Actual behavior

As can be seen in the log output, both the `CounterBean#increment()` and the `AlternativeCounterBean#increment()` are invoked by the scheduler.

> [INFO] Running org.acme.scheduler.CountResourceTest
2022-03-09 12:40:11,111 INFO  [org.jbo.threads] (main) JBoss Threads version 3.4.2.Final
2022-03-09 12:40:12,757 INFO  [io.quarkus] (main) Quarkus 2.7.3.Final on JVM started in 2.108s. Listening on: http://localhost:8081
2022-03-09 12:40:12,758 INFO  [io.quarkus] (main) Profile test activated. 
2022-03-09 12:40:12,758 INFO  [io.quarkus] (main) Installed features: [cdi, resteasy, scheduler, smallrye-context-propagation, vertx]
2022-03-09 12:40:13,011 INFO  [org.acm.sch.CounterBean] (executor-thread-1) Invoked from: org.acme.scheduler.CounterBean
2022-03-09 12:40:13,011 INFO  [org.acm.sch.CounterBean] (executor-thread-0) Invoked from: org.acme.scheduler.AlternativeCounterBean
[INFO] Tests run: 1, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 6.313 s - in org.acme.scheduler.CountResourceTest
2022-03-09 12:40:13,968 INFO  [io.quarkus] (main) Quarkus stopped in 0.048s

### How to Reproduce?

[scheduler-quickstart-reproducer.zip](https://github.com/quarkusio/quarkus/files/8216991/scheduler-quickstart-reproducer.zip)


### Output of `uname -a` or `ver`

Microsoft Windows [Version 10.0.19042.1526]

### Output of `java -version`

openjdk version "11.0.9.1" 2020-11-04 LTS OpenJDK Runtime Environment Corretto-11.0.9.12.1 (build 11.0.9.1+12-LTS) OpenJDK 64-Bit Server VM Corretto-11.0.9.12.1 (build 11.0.9.1+12-LTS, mixed mode)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.7.3.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.4 (9b656c72d54e5bacbed989b64718c159fe39b537) Maven home: C:\ProgramData\Chocolatey\lib\maven\apache-maven-3.8.4 Java version: 11.0.9.1, vendor: Amazon.com Inc., runtime: C:\Program Files\Amazon Corretto\jdk11.0.9_12 Default locale: en_US, platform encoding: Cp1252 OS name: "windows 10", version: "10.0", arch: "amd64", family: "windows"

### Additional information

_No response_

Thank you for your response, @mkouba.  Your recommendation for separating the logic and the scheduling in different beans provides us with a usable pattern.

> However, in your particular case I'd say that only the scheduled method declared on `CounterBean` should be executed because the overriding method declared on `AlternativeCounterBean` is not annotated with `@Scheduled`.

Would this mean that the behavior of the Quarkus scheduler needs to change?  At the moment in the code example I provided both the `AlternativeCounterBean` and `CounterBean` instances are having their `increment()` functions invoked by the scheduler.

This behavior seems to cause a problem if an identity is specified in the scheduled annotation on the `CounterBean`, like this: 
```
@Scheduled(every = "5s", identity = "counter_bean_identity")
void increment() {
    logger.info("Invoked from: {}", getClass().getCanonicalName());
    counter.incrementAndGet();
}
```
In that case with the `AlternativeCounterBean` defined like this, Quarkus fails to start with the following error:
```
@Alternative
@Priority(1)
@ApplicationScoped
public class AlternativeCounterBean extends CounterBean {
   
   @Override
   void increment() {
      super.increment();
   }
}
```
Error:

> 2022-03-10 08:36:06,934 ERROR [io.qua.dep.dev.IsolatedDevModeMain] (main) Failed to start quarkus: java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.arc.deployment.ArcProcessor#generateResources threw an exception: javax.enterprise.inject.spi.DeploymentException: java.lang.IllegalStateException: The identity: "counter_bean_identity" on: @Scheduled(every = "5s",identity = "counter_bean_identity") is not unique and it has already bean used by : @Scheduled(every = "5s",identity = "counter_bean_identity")
	at io.quarkus.arc.processor.BeanDeployment.processErrors(BeanDeployment.java:1202)
	at io.quarkus.arc.processor.BeanProcessor.processValidationErrors(BeanProcessor.java:147)
	at io.quarkus.arc.deployment.ArcProcessor.generateResources(ArcProcessor.java:492)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at io.quarkus.deployment.ExtensionLoader$2.execute(ExtensionLoader.java:882)
	at io.quarkus.builder.BuildContext.run(BuildContext.java:277)
	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at java.base/java.lang.Thread.run(Thread.java:829)
	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
Caused by: java.lang.IllegalStateException: The identity: "counter_bean_identity" on: @Scheduled(every = "5s",identity = "counter_bean_identity") is not unique and it has already bean used by : @Scheduled(every = "5s",identity = "counter_bean_identity")
	at io.quarkus.scheduler.deployment.SchedulerProcessor.validateScheduled(SchedulerProcessor.java:451)
	at io.quarkus.scheduler.deployment.SchedulerProcessor.validateScheduledBusinessMethods(SchedulerProcessor.java:200)
	... 11 more

If I don't use the explicitly defined `identity` and annotate the overriding method on `AlternativeCounterBean` with `@Scheduled` as well, there are three invocations, and if I understood your comment correctly it seems that there should be only two in that case:

> 2022-03-10 08:46:56,022 INFO  [org.acm.sch.CounterBean] (executor-thread-0) Invoked from: org.acme.scheduler.AlternativeCounterBean_Subclass
2022-03-10 08:46:56,021 INFO  [org.acm.sch.CounterBean] (executor-thread-2) Invoked from: org.acme.scheduler.CounterBean_Subclass
2022-03-10 08:46:56,022 INFO  [org.acm.sch.CounterBean] (executor-thread-1) Invoked from: org.acme.scheduler.AlternativeCounterBean_Subclass

Is this its intended behavior?

--- a/docs/src/main/asciidoc/scheduler-reference.adoc
+++ b/docs/src/main/asciidoc/scheduler-reference.adoc
@@ -33,8 +33,26 @@ Furthermore, the annotated method must return `void` and either declare no param
 
 TIP: The annotation is repeatable so a single method could be scheduled multiple times.
 
-TIP: A CDI event of type `io.quarkus.scheduler.SuccessfulExecution` is fired synchronously and asynchronously when an execution of a scheduled method is successful. A CDI event of type `io.quarkus.scheduler.FailedExecution` is fired synchronously and asynchronously  when an execution of a scheduled method throws an exception.
+[WARNING]
+====
+Subclasses never inherit the metadata of a `@Scheduled` method declared on a superclass. In the following example, the `everySecond()` method is only invoked upon the instance of `Jobs`.
+[source,java]
+----
+class Jobs {
 
+   @Scheduled(every = "1s")
+   void everySecond() {
+     // ..do something 
+   }
+}
+
+@Singleton
+class MyJobs extends Jobs {
+}
+----
+====
+
+A CDI event of type `io.quarkus.scheduler.SuccessfulExecution` is fired synchronously and asynchronously when an execution of a scheduled method is successful. A CDI event of type `io.quarkus.scheduler.FailedExecution` is fired synchronously and asynchronously  when an execution of a scheduled method throws an exception.
 
 === Triggers
 

--- a/extensions/scheduler/deployment/src/main/java/io/quarkus/scheduler/deployment/SchedulerProcessor.java
+++ b/extensions/scheduler/deployment/src/main/java/io/quarkus/scheduler/deployment/SchedulerProcessor.java
@@ -178,14 +178,6 @@ public class SchedulerProcessor {
                 LOGGER.debugf("Found scheduled business method %s declared on %s", method, bean);
             }
         }
-
-        DotName superClassName = beanClass.superName();
-        if (superClassName != null) {
-            ClassInfo superClass = index.getClassByName(superClassName);
-            if (superClass != null) {
-                collectScheduledMethods(index, transformedAnnotations, bean, superClass, scheduledBusinessMethods);
-            }
-        }
     }
 
     @BuildStep

--- a/extensions/scheduler/deployment/src/test/java/io/quarkus/scheduler/test/inheritance/ScheduledMethodNotInheritedTest.java
+++ b/extensions/scheduler/deployment/src/test/java/io/quarkus/scheduler/test/inheritance/ScheduledMethodNotInheritedTest.java
@@ -1,0 +1,47 @@
+package io.quarkus.scheduler.test.inheritance;
+
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+import javax.inject.Singleton;
+
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.scheduler.Scheduled;
+import io.quarkus.test.QuarkusUnitTest;
+
+public class ScheduledMethodNotInheritedTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest test = new QuarkusUnitTest()
+            .withApplicationRoot((jar) -> jar
+                    .addClasses(Jobs.class, MyJobs.class));
+
+    @Test
+    public void testExecution() throws InterruptedException {
+        assertTrue(Jobs.LATCH.await(5, TimeUnit.SECONDS));
+        assertTrue(Jobs.JOB_CLASSES.stream().allMatch(s -> s.equals(Jobs.class.getName())));
+    }
+
+    @Singleton
+    static class MyJobs extends Jobs {
+
+    }
+
+    static class Jobs {
+
+        static final CountDownLatch LATCH = new CountDownLatch(2);
+        static final List<String> JOB_CLASSES = new CopyOnWriteArrayList<>();
+
+        @Scheduled(every = "1s", identity = "foo")
+        void everySecond() {
+            JOB_CLASSES.add(getClass().getName());
+            LATCH.countDown();
+        }
+    }
+}

